### ISSUE TITLE
Version 6 - 2021.03.04 Upgrade - PIP not installed

### ISSUE THREAD

**Issue by mburns75** [tprelog/iocage-homeassistant#38]
I think I broke my install trying to upgrade to 2021.03.04

After entering:
     get-pip-required homeassistant

I get the following: 
     get-pip-required: Command not found.

Obviously, this prevents me from continuing with the install, and since the prior step was to rm -f the HomeAssistant install directory, I'm without a running instance.  I could start from scratch with a fresh install, but I'd rather not, unless absolutely necessary.

I believe the "Command not found" error means that pip is not installed in the HomeAssistant Plugin jail.  I've tried "pkg install pip" but pkg then complains:
     No packages available to install matching 'pip' have been found in the repositories

I'm hoping I can get back on track here if I knew the exact package name in which pip is included.  What's the pkg install command that I should be using to get pip added to the jail?

**Comment by tprelog** [comment]
Sorry for the trouble @mburns75 -- We should be able to get you back on track...

There's not a `pip` package installed anywhere outside the virtualenv... This is intentional to help prevent accidentally using pip before the venv is activated.

>get-pip-required: Command not found.

This is a script I wrote (not something that gets installed with `pkg`) The script should be located in `/root/.plugin/bin`

Please share the output from `ls -al /root/.plugin/bin`

If the script appears in the directory - can it run using the full path?

```
/root/.plugin/bin get-pip-required homeassistant
```

Finally, please also share the output from

```
sysrc -a | grep plugin
```

**Comment by tprelog** [comment]
Alternatively, you should also be able to manually set you plugin back to version 5, turn on the automatic reinstall, then run the Plugin UPDATE again

```
iocage console $JAIL_NAME
sysrc plugin_version=5.0
sysrc plugin_clean_install_service=yes
```

The run Plugin UPDATE again from TrueNAS UI

**Comment by mburns75** [comment]
Hey!  No apologies needed @tprelog !  Your script has saved me countless hours through the management of my HA install.  Thanks!  Besides, I enjoy working through some broken things from time to time.  How better to learn?

First, the following is the output of `ls -al /root/.plugin/bin`

> root@HomeAssistant:/ # ls -al /root/.plugin/bin/
> total 32
> drwxr-xr-x  2 root  wheel     7 Mar  7 08:40 .
> drwxr-xr-x  3 root  wheel     3 Mar  6 23:08 ..
> -rwxr-xr-x  1 root  wheel  1256 Mar  7 08:40 get-pip-required
> -rwxr-xr-x  1 root  wheel  1354 Mar  7 08:40 install-hacs
> -rwxr-xr-x  1 root  wheel  6281 Mar  7 08:40 menu
> -rwxr-xr-x  1 root  wheel  5864 Mar  7 08:40 menu-service
> -rwxr-xr-x  1 root  wheel   678 Mar  7 08:40 plugin-info

Second, I ran the script using the full path and it appears to have run without error.

Third, the following is the output from `sysrc -a | grep plugin`  (note, I ran this after running "get-pip-required" and prior to following the rest of the upgrade instructions.)

> root@HomeAssistant:~ # sysrc -a | grep plugin
> plugin_clean_install_service: YES
> plugin_ini: v_0.4.0_20200816
> plugin_ver: ver_0.4.0
> root@HomeAssistant:~ #

Lastly, the following is the output from `echo $path`

> root@HomeAssistant:~ # echo $path
> /sbin /bin /usr/sbin /usr/bin /usr/local/sbin /usr/local/bin /root/bin

I will now attempt to follow the rest of the upgrade instructions since `/root/.plugin/bin/get-pip-required homeassistant` ran successfully.  I'll report back shortly.

```

**Comment by tprelog** [comment]
Standing by for results.... but I see the first problem.

**Comment by mburns75** [comment]
The next step in the install instructions returned an error:

> root@HomeAssistant:~ # service homeassistant install homeassistant -r /root/.plugin/pip/requirements.homeassistant
> /usr/local/etc/rc.d/homeassistant: ERROR: /usr/local/share/homeassistant/bin/hass is not found or is not executable

So, I'm going to stop trying to upgrade manually and will instead use your suggestion of reverting back to plugin version 5 and then upgrading through the TrueNAS UI.

I'll report results shortly!

**Comment by tprelog** [comment]
WAIT!! STOP

**Comment by tprelog** [comment]
sorry meant to quickly comment, not hit close

**Comment by mburns75** [comment]
Yeah.  I saw your edit.  :)  I'll wait for further instruction.

**Comment by tprelog** [comment]
There's gonna be a lot to change, I'm sorry -- I can help you install a previous version of the plugin if you don't have time to deal with this...

But basically it looks like your still using the old `hass` user -- which has been switched to `homeassistant` -- Lets just confirm that real quick -- This will determine what's next

```
id 8123
```

**Comment by mburns75** [comment]
> root@HomeAssistant:~/.plugin/bin # id 8123
> uid=8123(homeassistant) gid=8123(homeassistant) groups=8123(homeassistant),68(dialer)

If this is end user error (me) don't feel you have to help me fix my own mistakes.  

I have previously installed and then uninstalled appdaemon.  I am using the file editor.

**Comment by mburns75** [comment]
> There's gonna be a lot to change, I'm sorry -- I can help you install a previous version of the plugin if you don't have time to deal with this...

Do you think I'm better off with a fresh install at this point? 

Still, I'm happy to help troubleshoot if that helps others.

**Comment by tprelog** [comment]
>If this is end user error (me) don't feel you have to help me fix my own mistakes.

Nope -- Not what's happening here -- Besides it helps me to have ideas to improve the plugin, when I am able to help.


>I have previously installed and then uninstalled appdaemon. I am using the file editor.

Cool - appdaemon is the tricky one

>uid=8123(homeassistant) gid=8123(homeassistant) groups=8123(homeassistant),68(dialer)

but it looks like you have the correct user :thinking: 

>Do you think I'm better off with a fresh install at this point?

This is the easy button -- Probably least amount of down time

>Still, I'm happy to help troubleshoot if that helps others.

It does -- If we can figure out where the update went wrong, I can try to fix it -- but I don't want to waste your time helping me fix my mistakes... so we can end this at any time

Please share

```
sysrc -a
```

```
ls -al /home
```

**Comment by mburns75** [comment]
You kidding me?  A chance to give back a bit for how much I've gotten out of this community?  I'm in.

```
root@HomeAssistant:~ # sysrc -a
configurator_config_dir: /home/homeassistant/configurator
configurator_enable: yes
configurator_group: homeassistant
configurator_python: /usr/local/bin/python3.7
configurator_umask: 002
configurator_user: homeassistant
configurator_venv: /usr/local/share/configurator
cron_flags:  -J 15
homeassistant_config_dir: /home/homeassistant/homeassistant
homeassistant_enable: yes
homeassistant_group: homeassistant
homeassistant_python: /usr/local/bin/python3.7
homeassistant_umask: 002
homeassistant_user: homeassistant
homeassistant_venv: /usr/local/share/homeassistant
hostname: HomeAssistant
ifconfig_epair0b: SYNCDHCP
ipv6_activate_all_interfaces: YES
plugin_clean_install_service: YES
plugin_ini: v_0.4.0_20200816
plugin_ver: ver_0.4.0
pubing_clean_install_service: YES
sendmail_enable: NO
sendmail_msp_queue_enable: NO
sendmail_outbound_enable: NO
sendmail_submit_enable: NO
syslogd_flags: -c -ss
```

```
root@HomeAssistant:~ # ls -al /home
total 18
drwxr-xr-x   3 root           wheel           3 Aug 16  2020 .
drwxr-xr-x  20 root           wheel          24 Oct 29 22:07 ..
drwxrwxr-x   6 homeassistant  homeassistant   7 Nov 24 00:10 homeassistant

```

**Comment by tprelog** [comment]
So what's going on here -- basically I tried to allow upgrades from version 4 (in version 5 I was blocking it) -- I wasn't thinking about the python version changing...
```
homeassistant_python: /usr/local/bin/python3.7
```

let's give this a quick try

```
sysrc homeassistant_python: /usr/local/bin/python3.8
service homeassistant install homeassistant -r /root/.plugin/pip/requirements.homeassistant
```

**Comment by mburns75** [comment]
Second command returned an error.  

```
root@HomeAssistant:~ # sysrc homeassistant_python=/usr/local/bin/python3.8
homeassistant_python: /usr/local/bin/python3.7 -> /usr/local/bin/python3.8
root@HomeAssistant:~ # service homeassistant install homeassistant -r /root/.plugin/pip/requirements.homeassistant
/usr/local/etc/rc.d/homeassistant: ERROR: /usr/local/share/homeassistant/bin/hass is not found or is not executable
```
Also:```

root@HomeAssistant:/usr/local/share/homeassistant/bin # ls -la
total 59
drwxr-xr-x  2 homeassistant  homeassistant    14 Mar  6 22:55 .
drwxrwxr-x  5 homeassistant  homeassistant     7 Mar  6 22:55 ..
-rw-r--r--  1 homeassistant  homeassistant  2228 Mar  6 22:55 activate
-rw-r--r--  1 homeassistant  homeassistant  1280 Mar  6 22:55 activate.csh
-rw-r--r--  1 homeassistant  homeassistant  2432 Mar  6 22:55 activate.fish
-rwxr-xr-x  1 homeassistant  homeassistant   258 Mar  6 22:55 easy_install
-rwxr-xr-x  1 homeassistant  homeassistant   258 Mar  6 22:55 easy_install-3.7
-rwxr-xr-x  1 homeassistant  homeassistant   249 Mar  6 22:55 pip
-rwxr-xr-x  1 homeassistant  homeassistant   249 Mar  6 22:55 pip3
-rwxr-xr-x  1 homeassistant  homeassistant   249 Mar  6 22:55 pip3.7
lrwxr-xr-x  1 homeassistant  homeassistant     9 Mar  6 22:55 python -> python3.7
lrwxr-xr-x  1 homeassistant  homeassistant     9 Mar  6 22:55 python3 -> python3.7
lrwxr-xr-x  1 homeassistant  homeassistant    24 Mar  6 22:55 python3.7 -> /usr/local/bin/python3.7
-rwxr-xr-x  1 homeassistant  homeassistant   236 Mar  6 22:55 wheel
```

**Comment by tprelog** [comment]
oh -- sorry I should have you remove the previous attempt :facepalm: 

We should be doing a clean install of the virtualenv here

Let's make sure we have the updated rc script -- I don' need to see your's but please verify it [Line 238](https://github.com/tprelog/iocage-homeassistant/blob/master/overlay/usr/local/etc/rc.d/homeassistant#L238)

```
cat /usr/local/etc/rc.d/homeassistant | sed -n '238p'
```

What your looking for is that begins with ` if [ "${1}" == "${name}" ]`

If not stop here, let me know --

If yes, please remove existing venv and try again

```
rm -rf /usr/local/share/homeassistant
service homeassistant install homeassistant -r /root/.plugin/pip/requirements.homeassistant
```

**Comment by mburns75** [comment]
Okay.  cat /usr/local ... did return the text string you indicated.

Both commands appear to have completed successfully!

**Comment by mburns75** [comment]
Is it now as simple as restarting the service?

**Comment by tprelog** [comment]
So next do
```
service homeassistant check_config
```

you should see it installing more stuff :crossed_fingers:

**Comment by tprelog** [comment]
You can just restart it but you won't see what's happening unless there is an error.... and you'll need to wait for things to finish installing before the UI is reachable...

when you use the `check_config` it makes homeassistant install the stuff without starting -- and allows you to see some out put -- if this completes, then homeassistant will start quicker after the reinstall because this stuff is already installed

either way should be fine though

**Comment by mburns75** [comment]
Looking good!  Here's the more stuff:


```
root@HomeAssistant:/usr/local/share/homeassistant/bin # service homeassistant check_config
INFO:homeassistant.util.package:Attempting install of colorlog==4.7.2
Testing configuration at /home/homeassistant/homeassistant
INFO:homeassistant.util.package:Attempting install of home-assistant-frontend==20210302.5
INFO:homeassistant.util.package:Attempting install of aiohttp_cors==0.7.0
INFO:homeassistant.util.package:Attempting install of pillow==8.1.1
INFO:homeassistant.util.package:Attempting install of paho-mqtt==1.5.1
INFO:homeassistant.util.package:Attempting install of python-join-api==0.0.6
INFO:homeassistant.util.package:Attempting install of onkyo-eiscp==1.2.7
INFO:homeassistant.util.package:Attempting install of scapy==2.4.4
INFO:homeassistant.util.package:Attempting install of distro==1.5.0
INFO:homeassistant.util.package:Attempting install of PyNaCl==1.3.0
INFO:homeassistant.util.package:Attempting install of emoji==1.2.0
INFO:homeassistant.util.package:Attempting install of defusedxml==0.6.0
INFO:homeassistant.util.package:Attempting install of netdisco==2.8.2
INFO:homeassistant.util.package:Attempting install of async-upnp-client==0.14.13
INFO:homeassistant.util.package:Attempting install of hass-nabucasa==0.41.0
INFO:homeassistant.util.package:Attempting install of sqlalchemy==1.3.23
INFO:homeassistant.util.package:Attempting install of pymodbus==2.3.0
INFO:homeassistant.util.package:Attempting install of mutagen==1.45.1
INFO:homeassistant.util.package:Attempting install of gTTS==2.2.2
INFO:homeassistant.util.package:Attempting install of pyisy==2.1.1
INFO:homeassistant.util.package:Attempting install of pyHS100==0.3.5.2
INFO:homeassistant.util.package:Attempting install of wakeonlan==1.1.6

```

**Comment by tprelog** [comment]
That's it !!! You'll know when there is an error

When that's done, you can start homeassistant
```
service homeassistant start
```

we still need to fix up some other things... for the plugin though

**Comment by mburns75** [comment]
And we're back!  `service homeassistant start` executed successfully!  Web interface is available and entities are beginning to populate back again!

Thank you so much!  

How can I help with the plugin?

**Comment by mburns75** [comment]
@tprelog 

I really appreciate your help in getting me back up and running.  If there is anything I can do to help test the plugin today, or at any time, please send me a note.  It would be good to give back.

Thanks so much!

Matt

**Comment by tprelog** [comment]
Thanks @mburns75 

I think there's a few more house keeping items we need to make sure your fixed up...

Do you wan to do these manually, or I can let you know when I get this fixed up in the plugin, then you can try running the Plugin UPDATE again to see if I get the remaining changes?

**Comment by mburns75** [comment]
I'm up and running, so we've fixed my immediate issue.  For my own personal use, I'm fine updating once the plugin is modified.  However, if it helps with the plugin development to use my case as a test/pilot, I'm happy to try going the manual route.

**Comment by tprelog** [comment]
Thanks -- I'll let you know when it's done - I think I'm gonna enable the virtualenv to rebuild because I know it will fail if not... the idea is that I don't want to break the plugin for someone who just pushes Plugin UPDATE... Moving forward this will also be used, for example (and in your case) when Python is updated.

**Comment by tprelog** [comment]
At your convenience... Please try running the Plugin UPDATE again from the TrueNAS UI

Note: This is going to rebuild the virtualenv again, so please allow extra time to reach the Home Assistant UI after the update

After the update please log into the jail's console to be sure the menu starts (The updated version should look like this)
![image](https://user-images.githubusercontent.com/35702532/110253606-3ac8ff00-7f59-11eb-9b7f-a0ab5c9f043f.png)

and please share `sysrc -a` again

**Comment by mburns75** [comment]
No luck and I fear this isn't what you were hoping for.  See attached screenshot of the error message I received from the TrueNAS UI when I attempted to update through the UI.

<img width="438" alt="Screenshot 2021-03-07 214245" src="https://user-images.githubusercontent.com/63862966/110267565-150a1d00-7f8e-11eb-98d9-20becb667e84.png">

To confirm, I received this error message after (1) selecting "Community" in the drop down in the Plugin section of TrueNAS, (2) clicking "Refresh Index", and then (3), clicking "Update" in the HomeAssistant plugin.  I confirm that I ensured that the checkboxes for "Update jail as well" and "Confirm" were both checked.

I did double check and can still get to the jail's console after the attempted update.  I am including the output of sysrc -a below, though I suspect this isn't helpful as the Update didn't go through.

```
root@HomeAssistant:~ # sysrc -a
configurator_config_dir: /home/homeassistant/configurator
configurator_enable: yes
configurator_group: homeassistant
configurator_python: /usr/local/bin/python3.7
configurator_umask: 002
configurator_user: homeassistant
configurator_venv: /usr/local/share/configurator
cron_flags:  -J 15
homeassistant_config_dir: /home/homeassistant/homeassistant
homeassistant_enable: yes
homeassistant_group: homeassistant
homeassistant_python: /usr/local/bin/python3.8
homeassistant_umask: 002
homeassistant_user: homeassistant
homeassistant_venv: /usr/local/share/homeassistant
hostname: HomeAssistant
ifconfig_epair0b: SYNCDHCP
ipv6_activate_all_interfaces: YES
plugin_clean_install_service: YES
plugin_ini: v_0.4.0_20200816
plugin_ver: ver_0.4.0
pubing_clean_install_service: YES
sendmail_enable: NO
sendmail_msp_queue_enable: NO
sendmail_outbound_enable: NO
sendmail_submit_enable: NO
syslogd_flags: -c -ss
```

**Comment by tprelog** [comment]
That helps... this directly because of problem is my indecisiveness (and why I just blocked any version 4 upgrades)

```
plugin_ini: v_0.4.0_20200816
plugin_ver: ver_0.4.0
```

I'm checking for `v_0.4.0` but should be checking for `ver_0.4.0` -- I'll fix that

**Comment by tprelog** [comment]
OK - Please try again, when you have time

**Comment by tprelog** [comment]
You can also remove these

- this one for sure
```
sysrc -x pubing_clean_install_service
```

- This one is OK to leave, but not really required for every Plugin UPDATE 
- I will trigger this during the UPDATE, when it's required
```
sysrc -x plugin_clean_install_service
```

**Comment by tprelog** [comment]
> OK - Please try again, when you have time

I thought about it... and I'm gonna try a more general approach - I pushed another commit, I think this one will do it

**Comment by mburns75** [comment]
It took!  You did it!  Successfully updated!

![Screenshot from 2021-03-08 09-28-18](https://user-images.githubusercontent.com/63862966/110357282-1e849b00-8009-11eb-94fe-f1a83879321a.png)

Note:  I'm still on 12.1-RELEASE-p13 for the Jail. I wasn't sure if you want me to re-run the following from the original instructions, so I've held off for the time being:

```
iocage fetch 12.2-RELEASE
iocage console $JAIL_NAME
# Press 0 to exit the menu
sysrc plugin_clean_install_service=YES
```

**Comment by tprelog** [comment]
We're getting somewhere now :smile: 
 
> I confirm that I ensured that the checkboxes for "Update jail as well" and "Confirm" were both checked....
> Note: I'm still on 12.1-RELEASE-p13 for the Jail.

ugh, I guess the "Update jail as well" does not do what I expected...

I wonder if I should just ask people to use the command line instead...

Please, give this a try...

```
iocage fetch 12.2-RELEASE
iocage upgrade $JAIL_NAME
```

**Comment by tprelog** [comment]
I'm working through this upgrade from v4 with another user on the forum as well...

I think these are the final steps to try - If your able, please make a final report back on success or failure

Since you originally installed the plugin on v4, please reset the version, to ensure all steps of the update are run. We also want to be sure to remove `plugin_clean_install_service` before the upgrade. (I'll trigger this automatically now)

```
iocage console $JAIL_NAME
# Press 0 to exit the menu
sysrc plugin_ver=ver_0.4.0
sysrc -x plugin_clean_install_service
```

Now exit the jail's console. The remaining commands are run from TrueNAS (_not_ inside the jail)
```
iocage fetch 12.2-RELEASE
iocage upgrade $JAIL_NAME
```

**Comment by mburns75** [comment]
Showing you as I go.  First, after entering the plugin jail called HomeAssistant:

```
root@HomeAssistant:~ # sysrc plugin_ver=ver_0.4.0
plugin_ver:  -> ver_0.4.0
root@HomeAssistant:~ # sysrc -x plugin_clean_install_service
```
Next, after exiting the jail, in TrueNAS:  `sudo iocage fetch 12.2-RELEASE`  (Note, the command line prompted me to

```
Type the number of the desired RELEASE
Press [Enter] to fetch the default selection: (12.2-RELEASE)
```
In my case, I entered "5" for 12.2-RELEASE, and the fetch appeared to complete successfully.  I then entered `sudo iocage upgrade HomeAssistant` 

I did see the "lua script failed" notice that you warned about, which oddly brought some comfort as I knew I was on the right path.  It finished with:
```
Running post_update.sh script
plugin_version: v6-20210307 -> v6-20210307

HomeAssistant successfully upgraded!
```
As I last check, I went back to the TrueNAS UI and found:

![Screenshot from 2021-03-08 17-58-06](https://user-images.githubusercontent.com/63862966/110393090-f82a2400-8037-11eb-9588-4afcda845b46.png)

You did it!  This is awesome!

Thanks so much for your help and the hand-holding you provided.  I stand ready to help further test if needed.
